src/coreclr/gc/gcdesc.h

```diff
@@ -186,7 +186,9 @@ class CGCDesc
         return PTR_CGCDescSeries(PTR_size_t(PTR_CGCDesc(this))-1)-1;
     }
 
-    // Returns number of immediate pointers this object has.
+    // Returns number of immediate pointers this object has. It should match the number of
+    // pointers enumerated by go_through_object_cl macro. The implementation shape has intentional
+    // similarity with the go_through_object_cl macro.
```

Comment:
This is helpful. I looked through `go_through_object_cl` for the collectable scenario and I can follow the `ContainsPointers()` branch. It seems the `Collectible()` is first rather than second but I don't think that matters. What I am still unsure about is the `NumOfPointers += 1;` below. I don't see the why but accept this is perhaps a lack of GC understanding here. Is there a succinct reason or is this intertwined with broader GC knowledge I am missing?

```C
#define go_through_object_cl(mt,o,size,parm,exp)                            \
{                                                                           \
    if (header(o)->Collectible())                                           \
    {                                                                       \
        uint8_t* class_obj = get_class_object (o);                             \
        uint8_t** parm = &class_obj;                                           \
        do {exp} while (false);                                             \
    }                                                                       \
    if (header(o)->ContainsPointers())                                      \
    {                                                                       \
        go_through_object_nostart(mt,o,size,parm,exp);                      \
    }                                                                       \
}
#endif //COLLECTIBLE_CLASS
```

Reply:
Each `{exp}` reports one pointer. `GetNumPointers` is meant to return how many times `go_through_object_cl` is going to do `{exp}`. 

`{exp}` is done once under the `if (header(o)->Collectible())` path in the macro. It explains the `NumOfPointers += 1;` under `if (pMT->Collectible())` in this function. Does it make more sense now?

Reply:
Ah yes. The do-while loop does match the logic now and I see why you changed the code to use that loop type. Thanks. All good here.